# Let's explore the dataset


In [1]:
# importing libraries
import pandas as pd 
import numpy as np 
import seaborn as sns 
import matplotlib.pyplot as plt 
import warnings 
from matplotlib import style
warnings.filterwarnings('ignore')
%matplotlib inline

In [2]:
df = pd.read_csv('dataset\cat_train.csv')
df.head()

id  bin_0  bin_1  bin_2 bin_3 bin_4 nom_0      nom_1    nom_2       nom_3  \
0   0    0.0    0.0    0.0     F     N   Red  Trapezoid  Hamster      Russia   
1   1    1.0    1.0    0.0     F     Y   Red       Star  Axolotl         NaN   
2   2    0.0    1.0    0.0     F     N   Red        NaN  Hamster      Canada   
3   3    NaN    0.0    0.0     F     N   Red     Circle  Hamster     Finland   
4   4    0.0    NaN    0.0     T     N   Red   Triangle  Hamster  Costa Rica   

   ...      nom_9 ord_0        ord_1     ord_2 ord_3 ord_4  ord_5  day month  \
0  ...  02e7c8990   3.0  Contributor       Hot     c     U     Pw  6.0   3.0   
1  ...  f37df64af   3.0  Grandmaster      Warm     e     X     pE  7.0   7.0   
2  ...        NaN   3.0          NaN  Freezing     n     P     eN  5.0   9.0   
3  ...  f9d456e57   1.0       Novice  Lava Hot     a     C    NaN  3.0   3.0   
4  ...  c5361037c   3.0  Grandmaster      Cold     h     C     OZ  5.0  12.0   

  target  
0      0  
1      0  
2      0  
3      0  
4      0  

[5 rows x 25 columns]

In [4]:
df.shape

(600000, 25)

In [5]:
df.info()

&lt;class &#39;pandas.core.frame.DataFrame&#39;&gt;
RangeIndex: 600000 entries, 0 to 599999
Data columns (total 25 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   id      600000 non-null  int64  
 1   bin_0   582106 non-null  float64
 2   bin_1   581997 non-null  float64
 3   bin_2   582070 non-null  float64
 4   bin_3   581986 non-null  object 
 5   bin_4   581953 non-null  object 
 6   nom_0   581748 non-null  object 
 7   nom_1   581844 non-null  object 
 8   nom_2   581965 non-null  object 
 9   nom_3   581879 non-null  object 
 10  nom_4   581965 non-null  object 
 11  nom_5   582222 non-null  object 
 12  nom_6   581869 non-null  object 
 13  nom_7   581997 non-null  object 
 14  nom_8   582245 non-null  object 
 15  nom_9   581927 non-null  object 
 16  ord_0   581712 non-null  float64
 17  ord_1   581959 non-null  object 
 18  ord_2   581925 non-null  object 
 19  ord_3   582084 non-null  object 
 20  ord_4   582070 non-null  object 
 

In [6]:
df.describe(include = 'all')

id          bin_0          bin_1          bin_2   bin_3  \
count   600000.000000  582106.000000  581997.000000  582070.000000  581986   
unique            NaN            NaN            NaN            NaN       2   
top               NaN            NaN            NaN            NaN       F   
freq              NaN            NaN            NaN            NaN  366212   
mean    299999.500000       0.092301       0.185532       0.278704     NaN   
std     173205.225094       0.289451       0.388729       0.448362     NaN   
min          0.000000       0.000000       0.000000       0.000000     NaN   
25%     149999.750000       0.000000       0.000000       0.000000     NaN   
50%     299999.500000       0.000000       0.000000       0.000000     NaN   
75%     449999.250000       0.000000       0.000000       1.000000     NaN   
max     599999.000000       1.000000       1.000000       1.000000     NaN   

         bin_4   nom_0     nom_1    nom_2   nom_3  ...      nom_9  \
count   581953  581748    581844   581965  581879  ...     581927   
unique       2       3         6        6       6  ...       2218   
top          N     Red  Triangle  Hamster   India  ...  8f3276a6e   
freq    312344  323286    164190   164897  164869  ...        565   
mean       NaN     NaN       NaN      NaN     NaN  ...        NaN   
std        NaN     NaN       NaN      NaN     NaN  ...        NaN   
min        NaN     NaN       NaN      NaN     NaN  ...        NaN   
25%        NaN     NaN       NaN      NaN     NaN  ...        NaN   
50%        NaN     NaN       NaN      NaN     NaN  ...        NaN   
75%        NaN     NaN       NaN      NaN     NaN  ...        NaN   
max        NaN     NaN       NaN      NaN     NaN  ...        NaN   

                ord_0   ord_1     ord_2   ord_3   ord_4   ord_5  \
count   581712.000000  581959    581925  582084  582070  582287   
unique            NaN       5         6      15      26     190   
top               NaN  Novice  Freezing       n       N      Fl   
freq              NaN  160597    142726   70982   39978   10562   
mean         1.948224     NaN       NaN     NaN     NaN     NaN   
std          0.853904     NaN       NaN     NaN     NaN     NaN   
min          1.000000     NaN       NaN     NaN     NaN     NaN   
25%          1.000000     NaN       NaN     NaN     NaN     NaN   
50%          2.000000     NaN       NaN     NaN     NaN     NaN   
75%          3.000000     NaN       NaN     NaN     NaN     NaN   
max          3.000000     NaN       NaN     NaN     NaN     NaN   

                  day          month         target  
count   582048.000000  582012.000000  600000.000000  
unique            NaN            NaN            NaN  
top               NaN            NaN            NaN  
freq              NaN            NaN            NaN  
mean         4.112767       6.371317       0.187205  
std          2.034430       3.458959       0.390076  
min          1.000000       1.000000       0.000000  
25%          2.000000       3.000000       0.000000  
50%          5.000000       6.000000       0.000000  
75%          6.000000       8.000000       0.000000  
max          7.000000      12.000000       1.000000  

[11 rows x 25 columns]

In [7]:
df.isnull().sum()

id            0
bin_0     17894
bin_1     18003
bin_2     17930
bin_3     18014
bin_4     18047
nom_0     18252
nom_1     18156
nom_2     18035
nom_3     18121
nom_4     18035
nom_5     17778
nom_6     18131
nom_7     18003
nom_8     17755
nom_9     18073
ord_0     18288
ord_1     18041
ord_2     18075
ord_3     17916
ord_4     17930
ord_5     17713
day       17952
month     17988
target        0
dtype: int64

The dataset consists of all the kinds of categorical variables
- Nominal
- Ordinal
- Cyclic
- Binary

In [8]:
df.target.value_counts()

0    487677
1    112323
Name: target, dtype: int64

It's a binary classification problem. 

The target is skewed, thus the best metric for this binary classification problem would be Area Under the ROC Curve (AUC) .  we can use precision and recall too ,  but AUC combines these two metrics. Thus we will be using AUC to evaluate the model that we build on this dataset

Overall,  there are:
- Five binary variables
- Ten nominal variables
- Six ordinal variables
- Two cyclic variables
- And a target variable 

let's look at "ord_2" feature in the dataset. it consist of six different categories:
- Freezing
- Warm 
- cold
- Boiling Hot
- Hot
- Lava Hot

To convert these categories to numbers . A simple way of doing this would be to create a dictionary that maps these values to numbers starting from  to N-1 , where N is the total number of categories in agiven feature

In [9]:
df.ord_2.unique()

array([&#39;Hot&#39;, &#39;Warm&#39;, &#39;Freezing&#39;, &#39;Lava Hot&#39;, &#39;Cold&#39;, &#39;Boiling Hot&#39;, nan],
      dtype=object)

In [10]:
mapping ={
    'Freezing' : 0,
    'Cold' : 1,
    'Warm' : 2,
    'Hot' : 3,
    'Boiling Hot': 4 ,
    'Lava Hot' : 5  
}

In [11]:
df.ord_2.value_counts()

Freezing       142726
Warm           124239
Cold            97822
Boiling Hot     84790
Hot             67508
Lava Hot        64840
Name: ord_2, dtype: int64

In [12]:
df.loc[ : , 'ord_2'] = df.ord_2.map(mapping)

In [13]:
df.ord_2.value_counts()

0.0    142726
2.0    124239
1.0     97822
4.0     84790
3.0     67508
5.0     64840
Name: ord_2, dtype: int64

This type of encoding of categorical variables is know as **Label Encoding**, i.e., we are encoding every category as a numerical label.
We can do the same by using LabelEncoder from scikit-learn.

In [15]:
from sklearn import preprocessing 

# read the data
df = pd.read_csv('dataset\cat_train.csv')

# fill NaN values in ord_2 column
df.loc[ :, 'ord_2'] = df.ord_2.fillna('NONE')

# initialize LabelEncoder
lbl_enc = preprocessing.LabelEncoder()

# fit label encoder and transform values on ord_2 column
# P>S: do not use this directly fit first , then transform
df.loc[ :, 'ord_2'] = lbl_enc.fit_transform(df.ord_2.values)

Since LabelEncoder from scikit-learn does not handle NaN values , and column ord_2 has NaN values in it we are using fillna from pandas
we can use this directly in many tree-based models:
- Decision trees
-  Random forest
- Extra trees
or any other kind of boosted trees model

This type of encoding cannot be used in linear models ,  support vector machines or neural networks as they expect data to be normalized or standardized.
For these types  of models , we can binarize the data.

- Freezing    --> 0 --> 0 0 0
- Warm        --> 1 --> 0 0 1
- Cold        --> 2 --> 0 1 0
- Boiling Hot --> 3 --> 0 1 1
- Hot         --> 4 --> 1 0 0
- Lava Hot    --> 5 --> 1 0 1

This is just converting the categories to number and then converting them to their binary representation. we are spliting one feature into three feature (in this case) feature (or columns) .  If we have more categories , we might end up spliting into a lot more columns.

It becomes easy to store lots of **binarized variables** like this is we store them in a sparse format.  A sparse format is nothing but a representaton or way of storing data in memory in which you do not store all the values but only the values that matter. In the case of binary variables described above , all that matters is where we have ones (1s)

it's difficult to imagine a format like this but should become clear with an example.

In [16]:
import numpy as np 
# create our example feature matrix
example = np.array([

     [0 , 0 , 1],
     [1, 0, 0],
     [1, 0, 1]
])
# print size in bytes
print(example.nbytes)

36


To represents this matrix only with ones would be some kind of dictionary method in whicch keys are indices of ros and columns and values is 1:

- (0, 2)  1
- (1, 0)  1
- (2, 0)  1
- (2, 2)  1

A notation like this will occupy much less memory because it has t store only four values (in this case).  The total memory used will be 8x4 = 32 bytes. Any numpy array can be converted to sparse matrix by simple python code.

In [17]:
import numpy as np 
from scipy import sparse

# create our example feature matrix
example = np.array([
    [0, 0, 1],
    [1, 0, 0],
    [1, 0, 1]
])
#convert numpy array to sparse CSR matrix
sparse_example = sparse.csr_matrix(example)

# print sie of this sparse matrix
print(sparse_example.data.nbytes)

16


In [18]:
print(
    sparse_example.data.nbytes +
    sparse_example.indptr.nbytes +
    sparse_example.indices.nbytes
)

48


Even though the sparse representation of binarized features takes much lless memory than its dense representation, there is another transformation for categorical variables that takes even less memory. This is know as ** One Hot Encoding** . 
One hot encoding is a binary encoding too in the sense that there are only two values , 0s and 1s . However , it must be noted that it's not a binary representation.

Whenever you get categorical variables, follow these simple steps:
- fill the NaN values (this is very important)
- convert them to integers by applying label encoding using LabelEncoder f scikit-learn or by using a mapping dictionary. IF YOU DIDN'T FILL UP NaN values with something, you might have to take care of them in this steps
- create one-hot encoding . you can skip binarization!
- go for modeling !


Handling NaN data in categorical feature is quite essential else you can get the infamous error from scikit-learn's LabelEncoder:

This simply means that when you are transforming tha test data, you have NaN , values in it . it;s because you forget to handle them during training. One simple way to handle NaN values would be to drop them. well, it;s simple but not ideal. there might be situation where most of your data has NaN values, and thus , you cannot drop rows/columns . Another way to handling NaN values is to treat them as a completely new category. This is the most preferred way of handling NaN values. And can be achieved in a very simple manner if you are using pandas.

In [19]:
df.ord_2.value_counts()

2    142726
6    124239
1     97822
0     84790
3     67508
4     64840
5     18075
Name: ord_2, dtype: int64

In [20]:
df.ord_2.fillna('NONE').value_counts()

2    142726
6    124239
1     97822
0     84790
3     67508
4     64840
5     18075
Name: ord_2, dtype: int64

Let's assume that ord_2 did not have any NaN values . we see that all categories in this column have a significant count . There are not "rare" categories ; i.e, categories which appear only a small percentage of the total number of samples. Now let's assume that you deployed this model which uses this column in production and when the model and project is live , you get a categories in ord_2 column that is not present in train . your model pipeline , in this case will throw  error and there is nothing you can do aboout it, then you must modify your model pipeline in production and include a new category to these six categories. This categories is known as the "rare" category . A rare category is a category which is not seen very often and can include many different categories. You can also try to "predict" the unknow category by using a nearest neighbour model

Let's see the value counts in "ord_4" column after filling NaN values:


In [21]:
df.ord_4.fillna('NONE').value_counts()

N       39978
P       37890
Y       36657
A       36633
R       33045
U       32897
M       32504
X       32347
C       32112
H       31189
Q       30145
T       29723
O       25610
B       25212
E       21871
K       21676
I       19805
NONE    17930
D       17284
F       16721
W        8268
Z        5790
S        4595
G        3404
V        3107
J        1950
L        1657
Name: ord_4, dtype: int64

we see that some values appear only a couple of thousand times , and some appear almost 40000 times . NaNs are also see a lot .

we can now define our criteria for calling a value "rare" . let's say the requirement for a vlaue being rare in this column is a count of less than 2000.  So , it seems , J and L can be marked as rare values. with pandas , it is quite easy to replace categories based on count thresold.

In [23]:
df.ord_4 = df.ord_4.fillna('NONE')
df.loc[
    df['ord_4'].value_counts()[df['ord_4']].values < 2000, 'ord_4'
] = 'RARE'

In [24]:
df.ord_4.value_counts()

N       39978
P       37890
Y       36657
A       36633
R       33045
U       32897
M       32504
X       32347
C       32112
H       31189
Q       30145
T       29723
O       25610
B       25212
E       21871
K       21676
I       19805
NONE    17930
D       17284
F       16721
W        8268
Z        5790
S        4595
RARE     3607
G        3404
V        3107
Name: ord_4, dtype: int64

In [25]:
df.ord_4.value_counts()['W']

8268

We say that whatever the value count for a certain category is less than 2000, replace it with rare. So now when it comes to test data , all the new unseen categories will be mapped to "RARE" and all missing values will be Mapped to "NONE".
This approach will also ensure that the model works in a live setting , even if you have a new categories .

In [3]:
df = pd.read_csv('dataset\cat_train_folds.csv')

In [30]:
df.kfold.value_counts()

4    120000
3    120000
2    120000
1    120000
0    120000
Name: kfold, dtype: int64

In [4]:
# we can check the target distribution per fold
df[df.kfold == 0].target.value_counts()

0    97536
1    22464
Name: target, dtype: int64

In [5]:
df[df.kfold == 1].target.value_counts()

0    97536
1    22464
Name: target, dtype: int64

we see that the target distribution is the same. this what we needed..
One of the simplest models we can build is by one hot encoding all the data and using logistic regression

when you run the pyhton code in powershell it will give you warnings to ignore these warnings use **-W ignore** i.e
**python -W ignore ohe_logres.py**

Now apply random forest in dataset , before applying random forest to the dataset we will use label encoder insted of one-hot encoder

Now let's change the dataset to another dataset with a lot of categorical variables. dataset is US adult census data. The dataset contains some feature , and our job is to predict the salary bracket. Let's take a look at this dataset .

In [12]:
cols = [
    'age' , 'workclass' , 'fnlwgt',
    'education' , 'education.num' , 'marital.status','occupation','relationship','race','sex','capital.gain','capital.loss' , 'hours.per.week' ,'native.country' , 'income'
]
df = pd.read_csv('dataset/adult.data')
df.head()

39          State-gov   77516   Bachelors   13        Never-married  \
0  50   Self-emp-not-inc   83311   Bachelors   13   Married-civ-spouse   
1  38            Private  215646     HS-grad    9             Divorced   
2  53            Private  234721        11th    7   Married-civ-spouse   
3  28            Private  338409   Bachelors   13   Married-civ-spouse   
4  37            Private  284582     Masters   14   Married-civ-spouse   

         Adm-clerical   Not-in-family   White     Male   2174   0   40  \
0     Exec-managerial         Husband   White     Male      0   0   13   
1   Handlers-cleaners   Not-in-family   White     Male      0   0   40   
2   Handlers-cleaners         Husband   Black     Male      0   0   40   
3      Prof-specialty            Wife   Black   Female      0   0   40   
4     Exec-managerial            Wife   White   Female      0   0   40   

    United-States   &lt;=50K  
0   United-States   &lt;=50K  
1   United-States   &lt;=50K  
2   United-States   &lt;=50K  
3            Cuba   &lt;=50K  
4   United-States   &lt;=50K

In [13]:
df = pd.DataFrame(df.values , columns= cols)
df.head()

age          workclass  fnlwgt   education education.num  \
0  50   Self-emp-not-inc   83311   Bachelors            13   
1  38            Private  215646     HS-grad             9   
2  53            Private  234721        11th             7   
3  28            Private  338409   Bachelors            13   
4  37            Private  284582     Masters            14   

        marital.status          occupation    relationship    race      sex  \
0   Married-civ-spouse     Exec-managerial         Husband   White     Male   
1             Divorced   Handlers-cleaners   Not-in-family   White     Male   
2   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
3   Married-civ-spouse      Prof-specialty            Wife   Black   Female   
4   Married-civ-spouse     Exec-managerial            Wife   White   Female   

  capital.gain capital.loss hours.per.week  native.country  income  
0            0            0             13   United-States   &lt;=50K  
1            0            0             40   United-States   &lt;=50K  
2            0            0             40   United-States   &lt;=50K  
3            0            0             40            Cuba   &lt;=50K  
4            0            0             40   United-States   &lt;=50K

In [8]:
df.shape

(32560, 15)

This dataset has the following columns:

- age
- workclass
- fnlwgt
- education
- education.num
- marital.status
- occupation
- relationship
- race
- sex
- capital.gain
- capital.loss
- hours.per.week
- native.country
- income

In [16]:
df.to_csv('dataset/adult.csv', index = False)

In [17]:
df.income.value_counts()

 &lt;=50K    24719
 &gt;50K      7841
Name: income, dtype: int64

We see there are 7841 instance with income greater than 50K USD. this is ~24% of the total number of samples. this will keep the evaluation metrics same i.e AUC

Before we start we will drop some columns, which are numerical, namely,
- fnlwgt
- age
- capital.gain
- capital.loss
- hours.per.week

In [57]:
df = pd.read_csv('dataset/adult_folds.csv')
df.head()

age          workclass  fnlwgt      education  education.num  \
0   30            Private  155781        HS-grad              9   
1   43   Self-emp-not-inc  176069        HS-grad              9   
2   22            Private  151105   Some-college             10   
3   25            Private  106889        HS-grad              9   
4   43            Private  241928        HS-grad              9   

        marital.status      occupation     relationship    race      sex  \
0   Married-civ-spouse    Craft-repair          Husband   White     Male   
1   Married-civ-spouse   Other-service             Wife   White   Female   
2        Never-married           Sales   Other-relative   White   Female   
3   Married-civ-spouse    Adm-clerical             Wife   White   Female   
4            Separated    Adm-clerical    Not-in-family   Black   Female   

   capital.gain  capital.loss  hours.per.week  native.country  income  kfold  
0             0             0              70   United-States   &lt;=50K      0  
1             0             0              16   United-States   &lt;=50K      0  
2             0             0              18   United-States   &lt;=50K      0  
3             0             0              30   United-States   &lt;=50K      0  
4             0             0              32   United-States   &lt;=50K      0

In [58]:
df.income.isnull().sum()

0

In [59]:
target_mapping = {" <=50K":0 ," >50K":1}
df.loc[:,'income'] = df['income'].map(target_mapping)

In [60]:
df.income.value_counts()

0    24719
1     7841
Name: income, dtype: int64

During mapping the "income" column just remember that there is **"One space"** before the string started 

In [56]:
df.head()

age          workclass  fnlwgt      education  education.num  \
0   30            Private  155781        HS-grad              9   
1   43   Self-emp-not-inc  176069        HS-grad              9   
2   22            Private  151105   Some-college             10   
3   25            Private  106889        HS-grad              9   
4   43            Private  241928        HS-grad              9   

        marital.status      occupation     relationship    race      sex  \
0   Married-civ-spouse    Craft-repair          Husband   White     Male   
1   Married-civ-spouse   Other-service             Wife   White   Female   
2        Never-married           Sales   Other-relative   White   Female   
3   Married-civ-spouse    Adm-clerical             Wife   White   Female   
4            Separated    Adm-clerical    Not-in-family   Black   Female   

   capital.gain  capital.loss  hours.per.week  native.country  income  kfold  
0             0             0              70   United-States     NaN      0  
1             0             0              16   United-States     NaN      0  
2             0             0              18   United-States     NaN      0  
3             0             0              30   United-States     NaN      0  
4             0             0              32   United-States     NaN      0

In [25]:
df.income.dtype

dtype(&#39;O&#39;)